src/main/java/com/squareup/phrase/ListPhrase.java

```diff
@@ -0,0 +1,296 @@
+   * @param finalElementPattern separator for final elements in lists with 3 or more elements
+   * @throws IllegalArgumentException if pattern contains any syntax errors.
+   */
+  public static ListPhrase from(Resources res, int twoElementPattern,
+      int nonFinalElementPattern, int finalElementPattern) {
+    return ListPhrase.from(Phrase.from(res, twoElementPattern),
+        Phrase.from(res, nonFinalElementPattern), Phrase.from(res, finalElementPattern));
+  }
+
+  /**
+   * Entry point into this API.
+   *
+   * @param pattern separator for all elements
+   * @throws IllegalArgumentException if pattern contains any syntax errors.
+   */
+  public static ListPhrase from(CharSequence pattern) {
+    return ListPhrase.from(pattern, pattern, pattern);
+  }
+
+  /**
+   * Entry point into this API.
+   *
+   * @param twoElementPattern separator for 2-element lists
+   * @param nonFinalElementPattern separator for non-final elements of lists with 3 or more elements
+   * @param finalElementPattern separator for final elements in lists with 3 or more elements
+   * @throws IllegalArgumentException if pattern contains any syntax errors.
+   */
+  public static ListPhrase from(CharSequence twoElementPattern, CharSequence nonFinalElementPattern,
+      CharSequence finalElementPattern) {
+    return ListPhrase.from(Phrase.from(twoElementPattern), Phrase.from(nonFinalElementPattern),
+        Phrase.from(finalElementPattern));
+  }
+
+  /**
+   * Entry point into this API.
+   *
+   * @param twoElementPattern separator for 2-element lists
+   * @param nonFinalElementPattern separator for non-final elements of lists with 3 or more elements
+   * @param finalElementPattern separator for final elements in lists with 3 or more elements
+   * @throws IllegalArgumentException if pattern contains any syntax errors.
+   */
+  public static ListPhrase from(Phrase twoElementPattern, Phrase nonFinalElementPattern,
+      Phrase finalElementPattern) {
+    return new ListPhrase(twoElementPattern, nonFinalElementPattern, finalElementPattern);
+  }
+
+  public interface Formatter<T> {
+    CharSequence format(T item);
+  }
+
+  private final Phrase twoElementPattern;
+  private final Phrase nonFinalElementPattern;
+  private final Phrase finalElementPattern;
+
+  private ListPhrase(Phrase twoElementPattern, Phrase nonFinalElementPattern,
+      Phrase finalElementPattern) {
+
+    this.twoElementPattern = validateListPattern("two-element", twoElementPattern);
+    this.nonFinalElementPattern = validateListPattern("non-final", nonFinalElementPattern);
+    this.finalElementPattern = validateListPattern("final", finalElementPattern);
+  }
+
+  public <T> CharSequence format(T... items) {
+    return format(Arrays.asList(items));
+  }
+
+  public <T> CharSequence format(Iterable<T> items) {
+    return format(items, null);
+  }
+
+  public <T> CharSequence format(Iterable<T> items, Formatter<T> formatter) {
+    return formatList(asList(items), formatter);
```

Comment:
Got rid of the extra list, instead we just traverse the iterable first to calculate the size, unless it implements Collection and therefore has a known size.